reactive/flannel.py

```diff
@@ -1,254 +1,218 @@
+               'connection_string': etcd.get_connection_string(),
+               'cert_path': ETCD_PATH}
+    render('flannel.service', '/lib/systemd/system/flannel.service', context)
+    set_state('flannel.service.installed')
+    remove_state('flannel.service.started')
+
+
+@when('config.changed.iface')
+def reconfigure_flannel_service():
+    ''' Handle interface configuration change. '''
+    remove_state('flannel.service.installed')
+
+
+@when('flannel.binaries.installed', 'flannel.etcd.credentials.installed',
+      'etcd.available')
+@when_not('flannel.network.configured')
+def configure_network(etcd):
+    ''' Store initial flannel data in etcd. '''
+    data = json.dumps({
+        'Network': config('cidr'),
+        'Backend': {
+            'Type': 'vxlan'
+        }
+    })
+    cmd = "etcdctl "
+    cmd += "--endpoint '{0}' ".format(etcd.get_connection_string())
+    cmd += "--cert-file {0} ".format(ETCD_CERT_PATH)
+    cmd += "--key-file {0} ".format(ETCD_KEY_PATH)
+    cmd += "--ca-file {0} ".format(ETCD_CA_PATH)
+    cmd += "set /coreos.com/network/config '{0}'".format(data)
+    check_call(split(cmd))
+    set_state('flannel.network.configured')
+    remove_state('flannel.service.started')
+
+
+@when('config.changed.cidr')
+def reconfigure_network():
+    ''' Trigger the network configuration method. '''
+    remove_state('flannel.network.configured')
+
+
+@when('flannel.binaries.installed', 'flannel.service.installed',
+      'flannel.network.configured')
+@when_not('flannel.service.started')
+def start_flannel_service():
+    ''' Start the flannel service. '''
+    status_set('maintenance', 'Starting flannel service.')
+    if service_running('flannel'):
+        service_restart('flannel')
     else:
-        client = etcd.Client(host=host, port=int(port), protocol='http')
+        service_start('flannel')
+    set_state('flannel.service.started')
 
-    data = {'Network': hookenv.config('cidr'), 'Backend': {'Type': 'vxlan'}}
-    json_data = json.dumps(data)
-    client.write('/coreos.com/network/config', json_data)
 
+@when('flannel.service.started', 'cni.is-worker')
+@when_not('flannel.cni.available')
+def set_available(cni):
+    ''' Indicate to the CNI provider that we're ready. '''
+    cni.set_available()
+    set_state('flannel.cni.available')
 
-def _ingest_network_config():
-    ''' When flannel configures itself on first boot, it generates an
-    environment file (subnet.env).
 
-    return a tuple of subnet, and interface mtu'''
-
-    if not os.path.isfile('/var/run/flannel/subnet.env'):
-        # The host has not fully started, and we have no file.
-        hookenv.log('Did not find expected file: /var/run/flannel/subnet.env')
-        return
+@when('flannel.binaries.installed')
+@when_not('flannel.version.set')
+def set_flannel_version():
+    ''' Surface the currently deployed version of flannel to Juju '''
+    status_set('maintenance', 'Setting flannel version.')
```

Comment:
This seems excessive to communicate to the user the charm is digging up the version. We can probably just let it `application-version-set` and be done with it.